kbase-extension/static/kbase/js/util/icon.js

```diff
@@ -1,150 +1,246 @@
-                    .addClass(circle_classes + ' kb-data-list-logo-shifted' + i)
-                    .css({'color': 'white'}));
-            }
-        }
-        // Assume there are CSS rules for levels of indent we care about..
-        if (indent > 0) {
-            $logo.addClass('kb-data-list-level1');
-        }
-        else if ($logo.hasClass('kb-data-list-level1')) {
-            $logo.removeClass('kb-data-list-level1');
-        }
+    function makeDataIcon($logo, type, stacked) {
+        const iconType = stacked ? 'data-stack' : 'data',
+            color = _.has(iconSpec.color_mapping, type)
+                ? iconSpec.color_mapping[type]
+                : logoColorLookup(type),
+            icon = _.has(iconSpec.data, type)
+                ? iconSpec.data[type]
+                : iconSpec.data.DEFAULT;
 
-        $logo.append($('<i>')
-            .addClass(circle_classes)
-            .css({'color': circle_color}));
-        // to avoid repetition, define the func. here that will
-        // add one set of icons
-        var addLogoFunc = function(fa_icon, $logo, cls) {
-            $logo.append($('<i>')
-                .addClass(fa_icon + ' fa-inverse fa-stack-1x ' + cls));
-        };
-        if (isCustomIcon(icon)) {
-            // add custom icons (more than 1 will look weird, though)
-            _.each(icon, function(cls) { addLogoFunc('icon', $logo, cls); });
-        } else {
-            // add stack of font-awesome icons
-            _.each(icon, function(cls) { addLogoFunc('fa', $logo, cls); });
-        }
+        const iconStr = _makeIcon(iconType, icon[0], 'circle', color, stacked);
+        $logo.append(iconStr);
         return $logo;
     }
 
-    /**
-     * Whether the stack of icons is using font-awesome
-     * or our own custom set.
-     *
-     * @param {list of str} iconList  Icon classes, from icons.json
-     * @returns {boolean}
-     */
-    function isCustomIcon(iconList) {
-        return (iconList.length > 0 && iconList[0].length > 4 &&
-            iconList[0].substring(0, 4) == 'icon');
-    }
-
     /**
      * Get color for data or method icon.
      * @param {string} type String that gets "hashed" into a color.
      * @returns {string} Color code
      */
     function logoColorLookup (type) {
-        var color = iconColorMapping[type];
-        if ( color === undefined) {
-            // fall back to primitive hack that just guesses
-            var code = 0;
-            for (var i = 0; i < type.length; code += type.charCodeAt(i++));
-            color = iconColors[code % iconColors.length];
-        }
+        let color;
+        // fall back to primitive hack that just guesses
+        let code = 0;
+        for (let i = 0; i < type.length; code += type.charCodeAt(i++));
+        color = iconSpec.colors[code % iconSpec.colors.length];
+        console.log(`guessed icon colour for ${type}; came up with ${color}. code: ${code}, length: ${iconSpec.colors.length}`);
```

Comment:
🛑 ☠️ 🛑 ☠️ 🛑 ☠️ 🛑

I couldn't work out what this code is supposed to do -- @briehl do you have any idea? (It was only three years ago that you wrote it!) I don't know if the closed `for` loop is deliberate or not. I'd like to replace it with something a bit clearer so if you have suggestions, fire away!

The alternative would be to give icons classes that include the `type`, and then put all the colour info into the scss files, with some default colour for types that don't have an assigned colour. That seems like a cleaner option.

Reply:
I largely maintained that code, I think. The short version of what happens there is that the for loop hashes the "type" string into an integer, then uses that to pick a color from the list. That's there for the case where there's a new type that gets created without an assigned color - this will pick a consistent color for it that'll be (probably) distinct from other types without an assigned icon / color.

You're right, though, it's kinda hacky. I'm ok with moving that info to scss files, now that we have them set up. That's probably where it would best belong anyway.

Reply:
OK, thanks. I'll fix that up in the next PR.